
syntax = "proto3";

package no.mechatronics.sfi.grpc_fmu;

option java_outer_classname = "FmiDefinitions";

message ModelReference {
    uint32 fmu_id = 1;
}

enum Causality {
    INPUT = 0;
    OUTPUT = 1;
    PARAMETER = 2;
    CALCULATED_PARAMETER = 3;
    LOCAL = 4;
    INDEPENDENT = 5;
}

enum Variability {
    CONSTANT =  0;
    FIXED = 1;
    CONTINUOUS = 2;
    DISCRETE = 3;
    TUNABLE = 4;
}

enum Initial {
    EXACT = 0;
    APPROX = 1;
    CALCULATED = 2;
}

message ScalarVariables {
    repeated ScalarVariable values = 1;
}

message Start {
    oneof value {
        int32 intValue = 1;
        double realValue = 2;
        string strValue = 3;
        bool boolValue = 4;
    }
}

message ScalarVariable {
    string name = 1;
    string description = 2;
    Initial initial = 3;
    Causality causality = 4;
    Variability variability = 5;
    Start start = 6;
}

message Status {
    int32 code = 1;
    string value = 2;
}

message InitRequest {
    uint32 fmu_id = 1;
    uint32 start = 2;
}

message StepRequest {
    uint32 fmu_id = 1;
    double dt = 2;
}

message TerminateRequest {
    uint32 fmu_id = 1;
}

message Int {
    int32 value = 1;
}

message IntWrite {
    uint32 fmu_id = 1;
    string varName = 2;
    int32 value = 3;
}

message IntList {
    repeated int32 values = 1;
}

message Real {
    double value = 1;
}

message RealWrite {
    uint32 fmu_id = 1;
    string varName = 2;
    double value = 3;
}

message RealList {
    repeated double values = 1;
}

message Str {
    string value = 1;
}

message StrWrite {
    uint32 fmu_id = 1;
    string varName = 2;
    string value = 3;
}

message StrList {
    repeated string values = 1;
}

message Bool {
    bool value = 1;
}

message BoolWrite {
    uint32 fmu_id = 1;
    string varName = 2;
    bool value = 3;
}

message BoolList {
    repeated bool values = 1;
}

message Var {
    oneof value {
        int32 intValue = 1;
        double realValue = 2;
        string strValue = 3;
        bool boolValue = 4;
    }
}

message VarRead {
    uint32 fmu_id = 1;
    string varName = 2;
}

message VarWrite {
    uint32 fmu_id = 1;
    string varName = 2;
    oneof value {
        int32 intValue = 5;
        double realValue = 6;
        string strValue = 7;
        bool boolValue = 8;
    }
}

message Empty {

}
